### ISSUE TITLE
Automatic report tailored to specific diverged cases

### ISSUE THREAD

**Issue by tomflexcompute** [flexcompute/tidy3d#1302]
Currently when a simulation diverged, an automatic email with the link to the [troubleshooting guide](https://www.flexcompute.com/tidy3d/examples/notebooks/DivergedFDTDSimulation/) will be sent to the user. This has been working to a certain extend, but some users still don't read it carefully and hence the divergence issue will not be resolved. 

Is it possible to automatically detect the potential cause of the divergence given the specific case and include that in the email? For example, if there is dispersive material into PML, the email can say "there is dispersive material into PML, which could potentially be the cause of divergence. Consider use absorber instead of PML." If there are structures close to PML, we can say "there are structures close to PML, consider move PML boundary further away." 

Having this could potentially save us quite a lot of time looking into their issues. Is there any technical difficulty in this workflow?

**Comment by tomflexcompute** [comment]
> One thing that I want to do soon is pipe all frontend warnings to a file which we can attach to that email. So even if the user disregarded a warning when building a simulation, they'd hopefully open that log once they get the diverged email. This should help with the structures close to PML issue, if it is caught by the warning. Currently we only check for 0.5lambda I believe, which might not catch all cases...
> 
> Once this is done we could try building up on that, e.g.
> 
> * [ ]  Automatically check if any divergence-related warning is in that log, and include it in the email
> * [ ]  We currently don't have a validator for dispersive materials in the PML. We probably don't want to have it on the frontend, because that could be fine in many cases. So we need to do a backend check about it and write to that log file. We also need to decide if we're doing a quick heuristic check (if any structures going into the PML are dispersive, even if overwritten by higher-up strucutres), or a proper (but slower) check where we actually build the structure containment and only then check for dispersive media.

Is it possible to check the dispersive medium or bend structure in PML only after the simulation diverged?

**Comment by momchil-flex** [comment]
Yeah we can do that. The translationally invariant check is even a bit more involved though. Although, @weiliangjin2021 maybe we can just put a permittivity monitor at the central frequency over the pml regions and judge base on that? It should be simple and it would probably capture all but some very exotic setups?

**Comment by weiliangjin2021** [comment]
> Yeah we can do that. The translationally invariant check is even a bit more involved though. Although, @weiliangjin2021 maybe we can just put a permittivity monitor at the central frequency over the pml regions and judge base on that? It should be simple and it would probably capture all but some very exotic setups?

Yeah, sounds straightforward.

**Comment by momchil-flex** [comment]
Merged in 2.6 backend, the new messages are written to `output/tidy3d_frontend.log`. @tomflexcompute as next step I think we should include that file in the divergence email, if the file exists.

I'm leaving the issue open since we should also automatically display the message in python e.g. when loading the simulation.

**Comment by tomflexcompute** [comment]
Thanks @momchil-flex . This is very nice and helpful. I'll work with MC to see if we can set up the automatic email with the file.

**Comment by momchil-flex** [comment]
Closing this in favor of #1483
